src/cls/rbd/cls_rbd_client.cc

```diff
@@ -2605,6 +2607,447 @@ int mirror_image_snapshot_set_copy_progress(librados::IoCtx *ioctx,
   return ioctx->operate(oid, &op);
 }
 
+void mirror_group_list_start(librados::ObjectReadOperation *op,
+                             const std::string &start, uint64_t max_return)
+{
+  bufferlist in_bl;
+  encode(start, in_bl);
+  encode(max_return, in_bl);
+  op->exec("rbd", "mirror_group_list", in_bl);
+}
+
+int mirror_group_list_finish(bufferlist::const_iterator *it,
+                             std::map<string, cls::rbd::MirrorGroup> *groups)
+{
+  try {
+    decode(*groups, *it);
+  } catch (const ceph::buffer::error &err) {
+    return -EBADMSG;
+  }
+  return 0;
+}
+
+int mirror_group_list(librados::IoCtx *ioctx,
+                      const std::string &start, uint64_t max_return,
+                      std::map<std::string, cls::rbd::MirrorGroup> *groups) {
+  librados::ObjectReadOperation op;
+  mirror_group_list_start(&op, start, max_return);
+
+  bufferlist out_bl;
+  int r = ioctx->operate(RBD_MIRRORING, &op, &out_bl);
+  if (r < 0) {
+    return r;
+  }
+
+  auto bl_it = out_bl.cbegin();
+  return mirror_group_list_finish(&bl_it, groups);
+}
+
+void mirror_group_resync_get_start(librados::ObjectReadOperation *op,
+                                   const std::string &global_group_id,
+                                   const std::string &group_name)
+{
+  bufferlist in_bl;
+  encode(global_group_id, in_bl);
+  encode(group_name, in_bl);
+  op->exec("rbd", "mirror_group_resync_get", in_bl);
+}
+
+int mirror_group_resync_get_finish(bufferlist::const_iterator *it,
+                                   std::string *group_id) {
+  try {
+    decode(*group_id, *it);
+  } catch (const ceph::buffer::error &err) {
+    return -EBADMSG;
+  }
+  return 0;
+}
+
+int mirror_group_resync_get(librados::IoCtx *ioctx,
+                            const std::string &oid,
+                            const std::string &global_group_id,
+                            const std::string &group_name,
+                            std::string *group_id)
+{
+  librados::ObjectReadOperation op;
+  mirror_group_resync_get_start(&op, global_group_id, group_name);
+
+  bufferlist out_bl;
+  int r = ioctx->operate(oid, &op, &out_bl);
+  if (r < 0) {
+    return r;
+  }
+
+  auto bl_it = out_bl.cbegin();
+  return mirror_group_resync_get_finish(&bl_it, group_id);
+}
+
+void mirror_group_resync_set(librados::ObjectWriteOperation *op,
+                             const std::string &global_group_id,
+                             const std::string &group_name,
+                             const std::string &group_id) {
```

Comment:
Writing all this is not necessary if it is using the group header. A simple key called say "resync" set to true should do. I'm also not convinced this is required to be a separate cls method if metadata_set could be used instead.

Reply:
The metadata_set code works. We do not need these functions at all as it just adds to the code and will need to be supported.